presto-main/src/main/java/com/facebook/presto/operator/project/DictionaryAwarePageProjection.java

```diff
@@ -0,0 +1,144 @@
+
+    private Block lastInputDictionary;
+    private Optional<Block> lastOutputDictionary;
+    private long lastDictionaryUsageCount;
+
+    public DictionaryAwarePageProjection(PageProjection projection, Function<DictionaryBlock, DictionaryId> sourceIdFunction)
+    {
+        this.projection = requireNonNull(projection, "projection is null");
+        this.sourceIdFunction = sourceIdFunction;
+        verify(projection.isDeterministic(), "projection must be deterministic");
+        verify(projection.getInputChannels().size() == 1, "projection must have only one input");
+    }
+
+    @Override
+    public Type getType()
+    {
+        return projection.getType();
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return projection.isDeterministic();
+    }
+
+    @Override
+    public InputChannels getInputChannels()
+    {
+        return projection.getInputChannels();
+    }
+
+    @Override
+    public Block project(ConnectorSession session, Page page, SelectedPositions selectedPositions)
+    {
+        Block block = page.getBlock(0);
+        if (block instanceof LazyBlock) {
+            block = ((LazyBlock) block).getBlock();
+        }
+
+        if (block instanceof RunLengthEncodedBlock) {
+            Block value = ((RunLengthEncodedBlock) block).getValue();
+            Optional<Block> projectedValue = processDictionary(session, value);
+            // single value block is always considered effective
+            verify(projectedValue.isPresent());
+            return new RunLengthEncodedBlock(projectedValue.get(), selectedPositions.size());
+        }
+
+        if (block instanceof DictionaryBlock) {
+            DictionaryBlock dictionaryBlock = (DictionaryBlock) block;
+            // Attempt to process the dictionary.  If dictionary is processing has not been considered effective, an empty response will be returned
+            Optional<Block> projectedDictionary = processDictionary(session, dictionaryBlock.getDictionary());
+            // record the usage count regardless of dictionary processing choice, so we have stats for next time
+            lastDictionaryUsageCount += selectedPositions.size();
+            // if dictionary was processed, produce a dictionary block; otherwise do normal processing
+            if (projectedDictionary.isPresent()) {
+                lastDictionaryUsageCount += selectedPositions.size();
+                int[] outputIds = filterDictionaryIds(dictionaryBlock, selectedPositions);
+                return new DictionaryBlock(selectedPositions.size(), projectedDictionary.get(), outputIds, false, sourceIdFunction.apply(dictionaryBlock));
+            }
+        }
+
+        return projection.project(session, new Page(block), selectedPositions);
+    }
+
+    private Optional<Block> processDictionary(ConnectorSession session, Block dictionary)
+    {
+        if (lastInputDictionary == dictionary) {
```

Comment:
Why the choice of caching just one (most recently used) dictionary.
Is that common that dictionary can be reused for subsequent pages but less probable to have a sequence of alternating dictionaries (A,B,A,B,A,B,A).

If latter is not less probable maybe we should replace `lastInputDictionary` with some size-bounded cache with LRU based eviction?